lib/dot-compat.js

```diff
@@ -0,0 +1,298 @@
+/**
+ * @fileoverview Compatibility class for dot config.
+ * @author Nicholas C. Zakas
+ */
+
+"use strict";
+
+//-----------------------------------------------------------------------------
+// Requirements
+//-----------------------------------------------------------------------------
+
+const path = require("path");
+const environments = require("../conf/environments");
+const createDebug = require("debug");
+const { ConfigArrayFactory } = require("./config-array-factory");
+
+//-----------------------------------------------------------------------------
+// Helpers
+//-----------------------------------------------------------------------------
+
+const debug = createDebug("eslintrc:dot-compat");
+const cafactory = Symbol("cafactory");
+
+/**
+ * Translates an ESLintRC-style config object into a dot-config-style config
+ * object.
+ * @param {Object} eslintrcConfig An ESLintRC-style config object.
+ * @param {Object} options Options to help translate the config.
+ * @param {string} options.resolveConfigRelativeTo To the directory to resolve
+ *      configs from.
+ * @param {string} options.resolvePluginsRelativeTo The directory to resolve
+ *      plugins from.
+ * @returns {Object} A dot-config-style config object.
+ */
+function translateESLintRC(eslintrcConfig, { resolveConfigRelativeTo, resolvePluginsRelativeTo }) {
+
+    const dotConfig = {};
+    const configs = [];
+    const languageOptions = {};
+    const linterOptions = {};
+    const keysToCopy = ["settings", "rules", "processor"];
+    const languageOptionsKeysToCopy = ["globals", "parser", "parserOptions"];
+    const linterOptionsKeysToCopy = ["noInlineConfig", "reportUnusedDisableDirectives"];
+    const pluginEnvironments = new Map();
+
+    // check for special settings for eslint:all and eslint:recommended:
+    if (eslintrcConfig.settings) {
+        if (eslintrcConfig.settings["eslint:all"] === true) {
+            return ["eslint:all"];
+        }
+
+        if (eslintrcConfig.settings["eslint:recommended"] === true) {
+            return ["eslint:recommended"];
+        }
+    }
+
+    // copy over simple translations
+    for (const key of keysToCopy) {
+        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {
+            dotConfig[key] = eslintrcConfig[key];
+        }
+    }
+
+    // copy over languageOptions
+    for (const key of languageOptionsKeysToCopy) {
+        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {
+            dotConfig.languageOptions = languageOptions;
+
+            if (languageOptions[key] && typeof languageOptions[key] === "object") {
```

Comment:
I'm having trouble figuring out the purpose of this branch. None of the tests cover it; the `else` is always taken. In what scenario would `languageOptions[key]` already be set?

Reply:
Hmmm, yeah, that looks like it's not doing what it should be doing. I think maybe `languageOptions` on that line should be `eslintrcConfig`, but I'll double-check.

Reply:
So it looks like my intent here was to create a clone of any object values instead of using the original object (for safety). `languageOptions` should be `eslintrcConfig`. I'll fix that.